test/cpp/src/020-heartbeat-listener.cc

```diff
@@ -0,0 +1,84 @@
+#include <catch.hpp>
+#include <typeinfo>
+#include <iostream>
+
+#include "listeners/heartbeat.h"
+#include "instruction.h"
+#include "trace_instruction.h"
+
+namespace
+{
+
+TEST_CASE("The heartbeat listener prints one line after 10M instructions retired") {
+    std::ostringstream stdout{};
+    Heartbeat uut{&stdout};
+    
+    // begin phase event
+    bool in_warmup = false;
+    uut.handle_event<Event::BEGIN_PHASE>(in_warmup);
+    
+    for (int i = 0; i < 5000000; ++i) {
+        std::deque<ooo_model_instr> fake_instructions{{ooo_model_instr(0, input_instr()), ooo_model_instr(0, input_instr())}};
+        uint32_t cpu = 0;
+        uint64_t curr_cycles = i;
+        auto cb = std::cbegin(fake_instructions);
+        auto ce = std::cend(fake_instructions);
+        uut.handle_event<Event::RETIRE>(cpu, cb, ce, curr_cycles);
+    }
+    
+    std::string res = stdout.str();
+    
+    REQUIRE(res.substr(0, res.find('(')) == "Heartbeat CPU 0 instructions: 10000000 cycles: 4999999 heartbeat IPC: 2 cumulative IPC: 2 ");

```

Comment:
Catch2 provides [matchers](https://github.com/catchorg/Catch2/blob/devel/docs/matchers.md#stdstring-matchers) that may give more detailed error messages here:

```suggestion
    REQUIRE_THAT(res, Catch::Matchers::StartsWith("Heartbeat CPU 0 instructions: 10000000 cycles: 4999999 heartbeat IPC: 2 cumulative IPC: 2 "));

```